In [1]:
import pandas as pd
import numpy as np

In [2]:
# read in the data
books = pd.read_csv("/Users/kapilojha/Desktop/ml/projects/book-recommender/data/books.csv",
                encoding="latin-1",
                on_bad_lines="skip",
                sep=";",
                low_memory=False
        )

In [3]:
!ls

research.ipynb


In [4]:
# view the data
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [5]:
print("Large URl: ", books["Image-URL-L"].iloc[0])
print("Medium URl: ", books["Image-URL-M"].iloc[0])
print("Small URl: ", books["Image-URL-S"].iloc[0])

Large URl:  http://images.amazon.com/images/P/0195153448.01.LZZZZZZZ.jpg
Medium URl:  http://images.amazon.com/images/P/0195153448.01.MZZZZZZZ.jpg
Small URl:  http://images.amazon.com/images/P/0195153448.01.THUMBZZZ.jpg


Each book has 3 diffrent Image url :
- depending on our requirements we can choose any of the three
- small, medium and large.

In [6]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271358 non-null  object
 3   Year-Of-Publication  271360 non-null  object
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: object(8)
memory usage: 16.6+ MB


In [7]:
books.shape

(271360, 8)

We have total of 2,71,360 rows and 8 features.

Since we do not need the small and medium url's we should remove them from
our dataset.

In [8]:
books.drop(["Image-URL-S", "Image-URL-M"],
           axis=1,
           inplace=True)
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [9]:
# rename the column names for easier processing
books.rename(columns={
    "ISBN" : "isbn",
    "Book-Title" : "title",
    "Book-Author" : "author",
    "Year-Of-Publication" : "year",
    "Publisher" : "publisher",
    "Image-URL-L" : "image-url"
}, inplace=True)
books.head()

isbn                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author  year                   publisher  \
0    Mark P. O. Morford  2002     Oxford University Press   
1  Richard Bruce Wright  2001       HarperFlamingo Canada   
2          Carlo D'Este  1991             HarperPerennial   
3      Gina Bari Kolata  1999        Farrar Straus Giroux   
4       E. J. W. Barber  1999  W. W. Norton &amp; Company   

                                           image-url  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [10]:
# import users data
users = pd.read_csv("/Users/kapilojha/Desktop/ml/projects/book-recommender/data/users.csv",
                    sep=";",
                    encoding="latin-1",
                    on_bad_lines="skip")
users["Age"] = users["Age"].bfill()
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa  18.0
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia  17.0
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom  61.0

In [11]:
# rename users columns for easier processing
users.rename(columns={
    "User-ID" : "id",
    "Location" : "location",
    "Age" : "age"
}, inplace=True)

In [12]:
users.head()

id                            location   age
0   1                  nyc, new york, usa  18.0
1   2           stockton, california, usa  18.0
2   3     moscow, yukon territory, russia  17.0
3   4           porto, v.n.gaia, portugal  17.0
4   5  farnborough, hants, united kingdom  61.0

In [13]:
# import the book ratings data
ratings = pd.read_csv("/Users/kapilojha/Desktop/ml/projects/book-recommender/data/ratings.csv",
                      sep=";",
                      encoding="latin-1",
                      on_bad_lines="skip")
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [14]:
# rename ratings columns names for easier processing
ratings.rename(columns={
    "User-ID" : "id",
    "ISBN" : "isbn",
    "Book-Rating" : "rating"
}, inplace=True)
ratings.head()

id        isbn  rating
0  276725  034545104X       0
1  276726  0155061224       5
2  276727  0446520802       0
3  276729  052165615X       3
4  276729  0521795028       6

In [15]:
# view the data shapes together
print(books.shape, users.shape, ratings.shape)

(271360, 6) (278858, 3) (1149780, 3)


In [16]:
# no. of books an individual user read
ratings["id"].value_counts()[:10]

id
11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
212898     4785
278418     4533
76352      3367
110973     3100
235105     3067
Name: count, dtype: int64

As you can see `user 11676` has read `13,602 books`.

In [17]:
# users who have at least read more than 200 books
x = ratings["id"].value_counts() > 200

In [18]:
x.value_counts()

count
False    104384
True        899
Name: count, dtype: int64

We can see only 899 users have read more than 200 books.

In [19]:
# extracting those users id's 
y = x[x].index
y

Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352, 110973,
       235105,
       ...
       260183,  73681,  44296, 155916,   9856, 274808,  28634,  59727, 268622,
       188951],
      dtype='int64', name='id', length=899)

In [20]:
# select only those rows with people who read more than 200 books in ratings dataset
ratings = ratings[ratings["id"].isin(y)]
ratings.head()

id        isbn  rating
1456  277427  002542730X      10
1457  277427  0026217457       0
1458  277427  003008685X       8
1459  277427  0030615321       0
1460  277427  0060002050       0

In [22]:
books.head()

isbn                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author  year                   publisher  \
0    Mark P. O. Morford  2002     Oxford University Press   
1  Richard Bruce Wright  2001       HarperFlamingo Canada   
2          Carlo D'Este  1991             HarperPerennial   
3      Gina Bari Kolata  1999        Farrar Straus Giroux   
4       E. J. W. Barber  1999  W. W. Norton &amp; Company   

                                           image-url  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [23]:
users.head()

id                            location   age
0   1                  nyc, new york, usa  18.0
1   2           stockton, california, usa  18.0
2   3     moscow, yukon territory, russia  17.0
3   4           porto, v.n.gaia, portugal  17.0
4   5  farnborough, hants, united kingdom  61.0

In [24]:
ratings.head()

id        isbn  rating
1456  277427  002542730X      10
1457  277427  0026217457       0
1458  277427  003008685X       8
1459  277427  0030615321       0
1460  277427  0060002050       0

In [25]:
# merge the ratings with books dataset using isbn number
ratings_with_books = ratings.merge(books, on="isbn")
ratings_with_books.head()

id        isbn  rating  \
0  277427  002542730X      10   
1  277427  0026217457       0   
2  277427  003008685X       8   
3  277427  0030615321       0   
4  277427  0060002050       0   

                                               title                 author  \
0  Politically Correct Bedtime Stories: Modern Ta...      James Finn Garner   
1                 Vegetarian Times Complete Cookbook             Lucy  Moll   
2                                           Pioneers  James Fenimore Cooper   
3   Ask for May, Settle for June (A Doonesbury book)          G. B. Trudeau   
4                  On a Wicked Dawn (Cynster Novels)      Stephanie Laurens   

   year                  publisher  \
0  1994  John Wiley &amp; Sons Inc   
1  1995      John Wiley &amp; Sons   
2  1974           Thomson Learning   
3  1982        Henry Holt &amp; Co   
4  2002                 Avon Books   

                                           image-url  
0  http://images.amazon.com/images/P/002542730X.0...  
1  http://images.amazon.com/images/P/0026217457.0...  
2  http://images.amazon.com/images/P/003008685X.0...  
3  http://images.amazon.com/images/P/0030615321.0...  
4  http://images.amazon.com/images/P/0060002050.0...

In [30]:
# create a dataframe with book title and no. of review of that titled book
number_of_ratings_per_book = ratings_with_books.groupby("title")["rating"].count().reset_index()
number_of_ratings_per_book.rename(columns={"rating" : "total_ratings"}, inplace=True)
number_of_ratings_per_book.head()

title  total_ratings
0   A Light in the Storm: The Civil War Diary of ...              2
1                              Always Have Popsicles              1
2               Apple Magic (The Collector's series)              1
3   Beyond IBM: Leadership Marketing and Finance ...              1
4   Clifford Visita El Hospital (Clifford El Gran...              1

In [31]:
# create a final dataset called book_ratings with the above dataset
book_ratings = number_of_ratings_per_book.merge(ratings_with_books, on="title")
book_ratings.head()

title  total_ratings      id  \
0   A Light in the Storm: The Civil War Diary of ...              2   35859   
1   A Light in the Storm: The Civil War Diary of ...              2   96448   
2                              Always Have Popsicles              1  172742   
3               Apple Magic (The Collector's series)              1  198711   
4   Beyond IBM: Leadership Marketing and Finance ...              1   11601   

         isbn  rating            author  year                    publisher  \
0  0590567330       0       Karen Hesse  1999  Hyperion Books for Children   
1  0590567330       9       Karen Hesse  1999  Hyperion Books for Children   
2  0964147726       0    Rebecca Harvin  1994            Rebecca L. Harvin   
3  0942320093       0  Martina Boudreau  1984           Amer Cooking Guild   
4  0962295701       0        Lou Mobley  1989       Teleonet, Incorporated   

                                           image-url  
0  http://images.amazon.com/images/P/0590567330.0...  
1  http://images.amazon.com/images/P/0590567330.0...  
2  http://images.amazon.com/images/P/0964147726.0...  
3  http://images.amazon.com/images/P/0942320093.0...  
4  http://images.amazon.com/images/P/0962295701.0...